File: LeetCode/java/1319_Number_of_Operations_to_Make_Network_Connected.java
import java.util.*;

class Solution {
    private void dfs(Map<Integer, List<Integer>> map, boolean[] visit, int prev, int cur, List<Pair> list) {
        if (prev != -1) {
            if (prev < cur) {
                list.add(new Pair(prev, cur));
            } else {
                list.add(new Pair(cur, prev));
            }
        }

        visit[cur] = true;

        List<Integer> current = map.get(cur);
        for (int i = 0; i < current.size(); i++) {
            if (visit[current.get(i)] == false) {
                dfs(map, visit, cur, current.get(i), list);
            }
        }
    }

    public int makeConnected(int n, int[][] connections) {
        Map<Integer, List<Integer>> map = new HashMap<>();
        boolean[] visit = new boolean[n];
        List<Pair> list = new ArrayList<>();
        Set<Pair> set = new HashSet<>();

        for (int[] con : connections) {
            if (!map.containsKey(con[0])) {
                map.put(con[0], new ArrayList<>());
            }

            map.get(con[0]).add(con[1]);

            if (!map.containsKey(con[1])) {
                map.put(con[1], new ArrayList<>());
            }

            map.get(con[1]).add(con[0]);
        }

        for (int[] con : connections) {
            set.add(new Pair(con[0], con[1]));
        }

        int connectionCnt = 0;
        for (int key : map.keySet()) {
            if (visit[key] == false) {
                dfs(map, visit, -1, key, list);
                connectionCnt++;
            }
        }

        for (Pair p : list) {
            set.remove(p);
        }

        int notVisitCnt = n - map.size();
        if (set.size() < notVisitCnt + connectionCnt - 1) {
            return -1;
        }

        return notVisitCnt + connectionCnt - 1;
    }
}

class Pair {
    int x;
    int y;

    Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }

    int getX() {
        return x;
    }

    int getY() {
        return y;
    }

    @Override
    public boolean equals(Object ob)
    {
        if (ob == this) {
            return true;
        }

        if (ob == null || ob.getClass() != getClass()) {
            return false;
        }

        Pair p = (Pair) ob;
        return Objects.equals(x, p.x) && p.y == y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    @Override
    public String toString() {
        return "{" + x + ", " + y + "}";
    }
}

File: LeetCode/cpp/1319_Number_of_Operations_to_Make_Network_Connected.cpp
class Solution {
public:
	void dfs(unordered_map<int, vector<int>>& um, vector<bool>& visit, int prev, int cur, vector<pair<int, int>>& vec) {
		if (prev != -1) {
			if (prev < cur) {
				vec.push_back({ prev, cur });
			}
			else {
				vec.push_back({ cur, prev });
			}
		}

		visit[cur] = true;

		vector<int> v = um[cur];
		for (int i = 0; i < v.size(); i++) {
			if (visit[v[i]] == false) {
				dfs(um, visit, cur, v[i], vec);
			}
		}
	}

	int makeConnected(int n, vector<vector<int>>& connections) {
		unordered_map<int, vector<int>> um;
		vector<bool> visit(n, false);
		vector<pair<int, int>> vec;
		set<pair<int, int>> st;

		for (vector<int> con : connections) {
			um[con[0]].push_back(con[1]);
			um[con[1]].push_back(con[0]);
		}

		for (vector<int> con : connections) {
			st.insert({ con[0], con[1] });
		}

		int connectionCnt = 0;
		for (auto m : um) {
			if (visit[m.first] == false) {
				dfs(um, visit, -1, m.first, vec);
				connectionCnt++;
			}
		}

		for (pair<int, int> v : vec) {
			st.erase(v);
		}

		int notVisitCnt = n - um.size();
		if (st.size() < notVisitCnt + connectionCnt - 1) {
			return -1;
		}

		return notVisitCnt + connectionCnt - 1;
	}
};